{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48c019ed-4e89-4c4a-9f9b-96f142dc9f35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleaned up manipulation of bio-mart exports\n",
    "# across multiple chromosomes\n",
    "\n",
    "# stephani needed more samples. this is all chromosomes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c2a3de6c-87db-4bb2-8c74-30cc65b833ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.Seq import Seq # for rev-comp\n",
    "from glob import glob\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import tempfile\n",
    "import subprocess\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad328d5a-f92f-4cfd-9cfe-e591e18a06c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def load_ref_chromosome(chrome_index: int,\n",
    "                        folder:str):\n",
    "    # get the path\n",
    "    ref_chrome_path = _get_ref_chromosome_path(chrome_index, folder)\n",
    "    # load the chrome\n",
    "    with open(ref_chrome_path,'r') as f:\n",
    "        ref_chrome = f.readlines()\n",
    "    # cleanup (drop name, just NTs)\n",
    "    ref_chrome = ''.join(ref_chrome[1:]).replace('\\n','')\n",
    "    return ref_chrome\n",
    "    \n",
    "def _get_ref_chromosome_path(chrome_index:int,\n",
    "                             folder:str):\n",
    "    ref_chrome_path = glob(f'{folder}*.{chrome_index}.fa')[0]\n",
    "    return ref_chrome_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8dbf2213-5b4c-4062-b84a-4d9ae8d671ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rev_comp(seq:str) -> str:\n",
    "    return Seq(seq).reverse_complement().__str__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a7b2284-1e75-4433-a156-484f7c1db276",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def remove_non_protein(df) -> pd.DataFrame:\n",
    "    df['Protein stable ID'] = df['Protein stable ID'].astype(str)\n",
    "    return df[df['Protein stable ID'] != 'nan']\n",
    "\n",
    "def strand_plus_minus(df) -> pd.DataFrame:\n",
    "    df['Strand'] = df['Strand'].replace(1,'+')\n",
    "    df['Strand'] = df['Strand'].replace(-1,'-')\n",
    "    return df\n",
    "\n",
    "def save_as_bed(df, bed_path):\n",
    "    ''' user must ensure correctness '''\n",
    "    # save\n",
    "    df.to_csv(bed_path,\n",
    "             header=None,\n",
    "             index=None,\n",
    "             sep='\\t')\n",
    "\n",
    "def load_bed(bed_path):\n",
    "    # load bed\n",
    "    z = pd.read_csv(bed_path,\n",
    "                    delimiter='\\t',\n",
    "                    header=None)\n",
    "    # rename columns for convenience\n",
    "    z = z.rename(columns = dict(zip(np.arange(6),\n",
    "                           ['chr','start','end','name','score','strand'])))\n",
    "    return z\n",
    "\n",
    "def save_fasta(save_path:str,\n",
    "               list_of_fastas: list[str]) -> None:\n",
    "    with open(save_path, 'w') as f:\n",
    "        for _line in list_of_fastas:\n",
    "            f.write(_line.replace('\\n','')+'\\n')\n",
    "\n",
    "def _getfasta(bed_path, genome_path):\n",
    "    # run the command\n",
    "    z = subprocess.run(args = ['./bedtools',\n",
    "                               'getfasta',\n",
    "                               '-s',\n",
    "                               '-fi',\n",
    "                               genome_path,\n",
    "                               '-bed',\n",
    "                               bed_path],\n",
    "                       cwd='/Users/Finn/bedtools/bin',\n",
    "                       capture_output=True\n",
    "                      )\n",
    "    # cleanup output into a fasta file\n",
    "    z = z.stdout.decode()\n",
    "    return z\n",
    "\n",
    "def _addslop(bed_path, genome_path,flank=10):\n",
    "    # load bed\n",
    "    temp_bed = load_bed(bed_path)\n",
    "    temp_bed['start'] -= flank\n",
    "    temp_bed['end'] += flank\n",
    "    with tempfile.NamedTemporaryFile() as f:\n",
    "        save_as_bed(temp_bed, bed_path = f.name)#\n",
    "        f.flush() # push the data \"in\" the file from buffer to disk\n",
    "        z = _getfasta(f.name, genome_path)\n",
    "\n",
    "    return z\n",
    "\n",
    "def save_fasta(save_path:str,\n",
    "               list_of_fastas: list[str]) -> None:\n",
    "    with open(save_path, 'w') as f:\n",
    "        for _line in list_of_fastas:\n",
    "            f.write(_line.replace('\\n','')+'\\n')\n",
    "            \n",
    "def export_to_bed(chrome, start, end, name, score, strand):\n",
    "    save_df = pd.DataFrame()\n",
    "    if isinstance(chrome, int):\n",
    "        save_df['chrome'] = np.ones(len(start),dtype=int)*chrome\n",
    "    else:\n",
    "        save_df['chrome'] = chrome\n",
    "    save_df['start'] = start # junction positions\n",
    "    save_df['end'] = end # junction positions +1\n",
    "    save_df['name'] = ['filler'] * len(start) #name.values # this doesn't transfer into \"getfasta\"\n",
    "    save_df['score'] = np.zeros(len(start)) # dumby scores\n",
    "    save_df['strand'] = strand\n",
    "    return save_df\n",
    "\n",
    "def load_mart(file_path: str) -> pd.DataFrame:\n",
    "    mart_export = pd.read_csv(file_path)\n",
    "    return mart_export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8adc291-3e4a-4406-9db5-8a3a7ffc7c41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeee8a58-2795-4055-956d-eccca803ec53",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a7aa39a0-3e24-4186-99db-9a3684ecce41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_junctions(mart_path:str, chrome_index:int,\n",
    "                  save_path:str\n",
    "                 ):\n",
    "    # load raw biomart file\n",
    "    temp_mart = load_mart(mart_path)\n",
    "    temp_mart = strand_plus_minus(temp_mart)\n",
    "    # build up bed files in dataframe\n",
    "    donors_start = list()\n",
    "    donors_end = list()\n",
    "    acceptors_start = list()\n",
    "    acceptors_end = list()\n",
    "    strand = list()\n",
    "\n",
    "    unique_genes = np.unique(temp_mart['Transcript stable ID version'])\n",
    "    # this isn't the right way to do this, should be a .groupby?\n",
    "    for _gene in unique_genes:\n",
    "        mask = (temp_mart['Transcript stable ID version'] == _gene)\n",
    "        temp_df = temp_mart[mask]\n",
    "        \n",
    "        if temp_df['Strand'].iloc[0] == '-':\n",
    "            # sort for safety\n",
    "            temp_df = temp_df.sort_values('Exon region start (bp)',ascending=False)\n",
    "            # add junctions\n",
    "            for i in temp_df['Exon region end (bp)'][1:]:\n",
    "                strand.append('-')\n",
    "                acceptors_start.append(i)\n",
    "                acceptors_end.append(i+1)\n",
    "            for i in temp_df['Exon region start (bp)'][:-1]:\n",
    "                donors_start.append(i-2)\n",
    "                donors_end.append(i-1)\n",
    "\n",
    "        else:\n",
    "            # sort for safety\n",
    "            temp_df = temp_df.sort_values('Exon region start (bp)')\n",
    "            # add junctions\n",
    "            for i in temp_df['Exon region end (bp)'][:-1]:\n",
    "                strand.append('+')\n",
    "                donors_start.append(i)\n",
    "                donors_end.append(i+1)\n",
    "            for i in temp_df['Exon region start (bp)'][1:]:\n",
    "                acceptors_start.append(i-2)\n",
    "                acceptors_end.append(i-1)\n",
    "            \n",
    "    # create bed table\n",
    "    junction_donor_bed = export_to_bed(chrome = chrome_index,\n",
    "                                       start = donors_start,\n",
    "                                       end = donors_end,\n",
    "                                       name = 'dumby_arg',\n",
    "                                       score = 'dumby_arg',\n",
    "                                       strand = strand)\n",
    "    save_as_bed(junction_donor_bed,\n",
    "                bed_path=f'{save_path}_donor.bed')\n",
    "    \n",
    "    junction_acceptor_bed = export_to_bed(chrome = chrome_index,\n",
    "                                       start = acceptors_start,\n",
    "                                       end = acceptors_end,\n",
    "                                       name = 'dumby_arg',\n",
    "                                       score = 'dumby_arg',\n",
    "                                       strand = strand)\n",
    "    save_as_bed(junction_acceptor_bed,\n",
    "                bed_path=f'{save_path}_acceptor.bed')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a209a547-59a9-40dc-9099-4d965a11eeb8",
   "metadata": {},
   "source": [
    "### split biomart output by chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3091ee98-a193-4c39-8cbc-3c8bbf25487a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split by chromosome\n",
    "# minimizing opening and closing of human chromosomes (and not requiring them to all be open at once)\n",
    "\n",
    "mart_path = '/Users/Finn/desktop/splicing_files/martquery_1119000515_815_all_chromes.txt'\n",
    "\n",
    "# cleanup\n",
    "temp_mart = load_mart(mart_path)\n",
    "temp_mart = strand_plus_minus(temp_mart)\n",
    "\n",
    "# split by chromosome number\n",
    "source = 'martquery_1119000515_815' # this helps keep track of the lineage of this data\n",
    "chromes = np.unique(temp_mart['Chromosome/scaffold name'])\n",
    "\n",
    "for _chrome in chromes:    \n",
    "    temp_mart[temp_mart['Chromosome/scaffold name'] == _chrome].to_csv(f'{raw_mart_folder}{source}_chr{_chrome}.txt',index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9546bda-8a1c-4944-b488-43149ca78840",
   "metadata": {},
   "source": [
    "#### loop over chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6c98d546-0aa3-49a1-b52f-e9eebe6f50e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_mart_folder = '/Users/Finn/desktop/splicing_files/mart_exports/'\n",
    "junction_folder = '/Users/Finn/desktop/splicing_files/junctions/'\n",
    "ref_genome_folder = '/Users/Finn/desktop/splicing_files/reference_chromes/'\n",
    "fasta_folder = '/Users/Finn/desktop/splicing_files/junction_fastas/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6f9461aa-0851-4d89-be9c-889b2b413ae3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on: 4\n",
      "working on: 5\n",
      "working on: 7\n",
      "working on: 6\n",
      "working on: 2\n",
      "working on: 3\n",
      "working on: 1\n",
      "working on: 13\n",
      "working on: 12\n",
      "working on: 10\n",
      "working on: 11\n",
      "working on: 15\n",
      "working on: 14\n",
      "working on: 16\n",
      "working on: 17\n",
      "working on: 19\n",
      "working on: 18\n",
      "working on: 20\n",
      "working on: 21\n",
      "working on: 22\n",
      "working on: 8\n",
      "working on: 9\n"
     ]
    }
   ],
   "source": [
    "flank_size = 100\n",
    "\n",
    "for _file in glob(f'{raw_mart_folder}*'):\n",
    "    # extract chrome index\n",
    "    _chrome = int(_file.split('/')[-1].split('chr')[1].split('.')[0])\n",
    "    print('working on:',_chrome)\n",
    "    \n",
    "    # extract junctions into bed files\n",
    "    get_junctions(mart_path = _file,\n",
    "                  chrome_index = _chrome,\n",
    "                  save_path = f'{junction_folder}chr{_chrome}')\n",
    "    \n",
    "    # extract acceptors\n",
    "    ref_chrome_path = glob(f'{ref_genome_folder}*.{_chrome}.fa')[0]\n",
    "    flanked = _addslop(f'{junction_folder}chr{_chrome}_acceptor.bed',\n",
    "                       ref_chrome_path, flank=flank_size).split('\\n')\n",
    "    save_fasta(f'{fasta_folder}chr{_chrome}_acceptor.fa', list_of_fastas=flanked)\n",
    "    \n",
    "    # extract donors\n",
    "    ref_chrome_path = glob(f'{ref_genome_folder}*.{_chrome}.fa')[0]\n",
    "    flanked = _addslop(f'{junction_folder}chr{_chrome}_donor.bed',\n",
    "                       ref_chrome_path, flank=flank_size).split('\\n')\n",
    "    save_fasta(f'{fasta_folder}chr{_chrome}_donor.fa', list_of_fastas=flanked)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08d012d4-db18-4a75-b61d-807c629da6ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "59d375d2-d5bd-4c40-acdd-66f16a6b585a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mart_path = '/Users/Finn/desktop/splicing_files/mart_exports/martquery_1108044906_870_chr19.txt'\n",
    "ref_genome_folder = '/Users/Finn/desktop/splicing_files/reference_chromes/'\n",
    "\n",
    "d19_bed = collect_negative_junctions(mart_path = mart_path,\n",
    "                            chrome_index = 19,\n",
    "                            ref_genome_folder = ref_genome_folder,\n",
    "                            junction_type = 'D')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "7ebf97ff-80e9-4909-8929-6bb7117634f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_as_bed(d19_bed,\n",
    "            bed_path = '/Users/Finn/desktop/d19_contol.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "589df6ae-6d5f-4798-8210-69676a922f83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>19:46022941-46022946(+)',\n",
       " 'GGGTC',\n",
       " '>19:46019591-46019596(+)',\n",
       " 'CCGTG',\n",
       " '>19:46019183-46019188(+)',\n",
       " 'CAGTG',\n",
       " '>19:1285894-1285899(+)',\n",
       " 'GGGTG',\n",
       " '>19:1295582-1295587(+)',\n",
       " 'CGGTG']"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_fasta = _addslop(bed_path = '/Users/Finn/desktop/d19_contol.bed',\n",
    "         genome_path=_get_ref_chromosome_path(chrome_index=19,\n",
    "                                              folder=ref_genome_folder),\n",
    "         flank=2).split('\\n')\n",
    "\n",
    "control_fasta[:10] # looks good"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7abcba3a-2980-4a60-bf74-f0e21b1c344e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "85ff9018-f857-4d98-8e91-2c91814edc8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "a19_bed = collect_negative_junctions(mart_path = mart_path,\n",
    "                            chrome_index = 19,\n",
    "                            ref_genome_folder = ref_genome_folder,\n",
    "                            junction_type = 'A')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "99b03b9c-ef32-46b3-b378-7391ab9c632b",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_as_bed(a19_bed,\n",
    "            bed_path = '/Users/Finn/desktop/a19_contol.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "0cb0b81a-0cf8-4ae9-99c6-4415af5b9780",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>19:46022985-46022990(+)',\n",
       " 'GAGAG',\n",
       " '>19:46019640-46019645(+)',\n",
       " 'CAGGA',\n",
       " '>19:46019179-46019184(+)',\n",
       " 'GAGAC',\n",
       " '>19:1286305-1286310(+)',\n",
       " 'CAGGT',\n",
       " '>19:1295794-1295799(+)',\n",
       " 'CAGCT',\n",
       " '>19:1298594-1298599(+)',\n",
       " 'AAGGT',\n",
       " '>19:1299837-1299842(+)',\n",
       " 'GAGGC',\n",
       " '>19:49072373-49072378(+)',\n",
       " 'TAGTA',\n",
       " '>19:49070488-49070493(+)',\n",
       " 'GAGGA',\n",
       " '>19:39738010-39738015(+)',\n",
       " 'GAGCT']"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_fasta = _addslop(bed_path = '/Users/Finn/desktop/a19_contol.bed',\n",
    "         genome_path=_get_ref_chromosome_path(chrome_index=19,\n",
    "                                              folder=ref_genome_folder),\n",
    "         flank=2).split('\\n')\n",
    "\n",
    "control_fasta[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "866ece54-0937-43a8-8c07-02056a612e63",
   "metadata": {},
   "source": [
    "#### seems finalized\n",
    "how many exons do these single transcripts have?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e9b80e3c-5e44-4cc8-802a-809f113d3858",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "151009"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mart_path = '/Users/Finn/desktop/splicing_files/martquery_1114011419_771_all_chromes.txt'\n",
    "temp_mart = load_mart(mart_path)\n",
    "n_exons = list()\n",
    "unique_genes = np.unique(temp_mart['Transcript stable ID version'])\n",
    "len(unique_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d78d8cab-0a8f-4a1b-901e-7e7fa48b9690",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#mart_path = '/Users/Finn/desktop/splicing_files/mart_exports/martquery_1108044906_870_chr19.txt'\n",
    "mart_path = '/Users/Finn/desktop/splicing_files/martquery_1113195942_714_all_chromes.txt'\n",
    "temp_mart = load_mart(mart_path)\n",
    "\n",
    "n_exons = list()\n",
    "unique_genes = np.unique(temp_mart['Transcript stable ID version'])\n",
    "for _gene in unique_genes:\n",
    "    mask = (temp_mart['Transcript stable ID version'] == _gene)\n",
    "    temp_df = temp_mart[mask]\n",
    "    n_exons.append(len(temp_df))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "13840e94-ce94-4ac2-99c0-f88208b73c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41582"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ad281929-2a05-42a8-b3d9-f2e757dc72df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtMAAAEGCAYAAACn7xkwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8ekN5oAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVqUlEQVR4nO3df7BmdX0f8PfHJcSocf3Bjq386JIsY8PYVuIdpLWTOlZTCK6kNklhYmOUuqFTjOkk0246naTt1BnSprWoRGdVJKYWQtHqrtBCJtXQJmoAf0R+1ClDMCxVQdSrNY0E/fSP+1BuFnb32bP32eecu6/XzJ37nO/zPOf53D1z7r73u5/zPdXdAQAAjtyTll0AAABMlTANAAADCdMAADCQMA0AAAMJ0wAAMNAJyy7gaJx00km9ffv2ZZcBAMAmd9ttt325u7cdOD7pML19+/bceuutyy4DAIBNrqo+/0Tj2jwAAGAgYRoAAAYSpgEAYKBJhumq2llVe1ZXV5ddCgAAx7FJhunu3tfdu7Zu3brsUgAAOI5NMkwDAMAYCNMAADCQMA0AAANN+qYty7J99/WPG7v3svOXUAkAAMtkZhoAAAYSpgEAYKBJhmnrTAMAMAaTDNPWmQYAYAwmGaYBAGAMhGkAABhImAYAgIGEaQAAGEiYBgCAgYRpAAAYSJgGAICBhGkAABhoVGG6qp5aVbdW1SuWXQsAABzOCYvceVVdmeQVSR7o7uevGz83yeVJtiR5V3dfNnvqnyS5dpE1LdL23dc/buzey85fQiUAABwLi56ZvirJuesHqmpLkiuSnJfkzCQXVdWZVfXyJHcmeWDBNQEAwIZY6Mx0d99cVdsPGD47yd3dfU+SVNU1SS5I8rQkT81awP6/VXVDd39nkfUBAMDRWGiYPoiTk9y3bnt/khd196VJUlU/neTLBwvSVbUrya4kOe200xZbKQAAHMKoLkBMku6+qrs/fIjn93T3SnevbNu27ViWBgAAf8YywvT9SU5dt33KbGxuVbWzqvasrq5uaGEAAHAklhGmb0lyRlWdXlUnJrkwyd4j2UF37+vuXVu3bl1IgQAAMI+FhumqujrJx5I8r6r2V9XF3f1IkkuT3JjkriTXdvcdR7hfM9MAACzdolfzuOgg4zckueEo9rsvyb6VlZXXD90HAAAcrdFdgAgAAFMxyTCtzQMAgDGYZJh2ASIAAGMwyTANAABjMMkwrc0DAIAxmGSY1uYBAMAYTDJMAwDAGEwyTGvzAABgDCYZprV5AAAwBpMM0wAAMAbCNAAADCRMAwDAQJMM0y5ABABgDCYZpl2ACADAGJyw7AKOB9t3X/+4sXsvO38JlQAAsJEmOTMNAABjIEwDAMBAkwzTLkAEAGAMJhmmXYAIAMAYTDJMAwDAGAjTAAAwkDANAAADCdMAADCQMA0AAAMJ0wAAMNAkw7R1pgEAGINJhmnrTAMAMAaTDNMAADAGwjQAAAwkTAMAwEDCNAAADCRMAwDAQMI0AAAMJEwDAMBAJyy7gOPZ9t3XP27s3svOX0IlAAAMYWYaAAAGGk2YrqofqKp3VNV1VfUPll0PAAAczkLDdFVdWVUPVNXtB4yfW1Wfq6q7q2p3knT3Xd19SZKfSPLiRdYFAAAbYdEz01clOXf9QFVtSXJFkvOSnJnkoqo6c/bcK5Ncn+SGBdcFAABHbaFhurtvTvKVA4bPTnJ3d9/T3Q8nuSbJBbPX7+3u85L85MH2WVW7qurWqrr1wQcfXFTpAABwWMtYzePkJPet296f5EVV9ZIkr0ry3TnEzHR370myJ0lWVlZ6YVUCAMBhjGZpvO7+aJKPzvPaqtqZZOeOHTsWWRIAABzSMlbzuD/Jqeu2T5mNza2793X3rq1bt25oYQAAcCSWEaZvSXJGVZ1eVScmuTDJ3iPZQVXtrKo9q6urCykQAADmseil8a5O8rEkz6uq/VV1cXc/kuTSJDcmuSvJtd19x5Hs18w0AABjsNCe6e6+6CDjN8TydwAATNxo7oB4JLR5AAAwBpMM09o8AAAYg9Esjcdjtu++/nFj9152/hIqAQDgUCY5M63NAwCAMZhkmNbmAQDAGEwyTAMAwBgI0wAAMNAkw7SeaQAAxmCSYVrPNAAAYzDJMA0AAGMgTAMAwECTDNN6pgEAGINJhmk90wAAjMEkwzQAAIyBMA0AAAMJ0wAAMJAwDQAAA50wz4uq6o1J3pPkG0neleSsJLu7+6YF1naoenYm2bljx45lfPzSbN99/ePG7r3s/CVUAgBAMv/M9Ou6++tJfjjJM5P8vSSXLayqw7CaBwAAYzBvmK7Z9x9J8hvdfce6MQAAOC7NG6Zvq6qbshamb6yq703yncWVBQAA4zdXz3SSi5O8IMk93f3HVfXsJK9dWFUAADAB885M/1Z3f7K7v5Yk3f1QkjcvrCoAAJiAQ85MV9WTkzwlyUlV9cw81if99CQnL7g2AAAYtcO1efxMkp9L8twkt+WxMP31JG9bXFkcCUvmAQAsxyHDdHdfnuTyqnpDd7/1GNV0WMfrOtMAAIzLXBcgdvdbq+qvJdm+/j3d/d4F1XW4evYl2beysvL6ZXw+AAAk898B8TeSfH+STyf59my4kywlTAMAwBjMuzTeSpIzu7sXWQwbSy81AMBizbs03u1J/twiCwEAgKmZd2b6pCR3VtXvJ/nWo4Pd/cqFVAUAABMwb5j+54ssAgAApmje1Tx+Z9GFAADA1My7msc3srZ6R5KcmOS7knyzu5++qMJYHBcmAgBsjHlnpr/30cdVVUkuSHLOoooCAIApmHc1j/+v13wwyd/a6GKq6ker6p1V9ZtV9cMbvX8AANhI87Z5vGrd5pOytu70n8z53iuTvCLJA939/HXj5ya5PMmWJO/q7stmIf2DVfXMJL+a5KZ5PgMAAJZh3tU8dq57/EiSe7PW6jGPq5K8LevullhVW5JckeTlSfYnuaWq9nb3nbOX/LPZ8wAAMFrz9ky/dugHdPfNVbX9gOGzk9zd3fckSVVdk+SCqroryWVJ/kt3f/KJ9ldVu5LsSpLTTjttaFkAAHDU5uqZrqpTquo/V9UDs6/3V9UpR/G5Jye5b932/tnYG5K8LMmPVdUlT/TG7t7T3SvdvbJt27ajKAEAAI7OvBcgvifJ3iTPnX3tm41tqO5+S3e/sLsv6e53HOx1VbWzqvasrq5udAkAADC3ecP0tu5+T3c/Mvu6KsnRTAvfn+TUddunzMbm0t37unvX1q1bj6IEAAA4OvOG6Yeq6tVVtWX29eokDx3F596S5IyqOr2qTkxyYdZmvgEAYDLmXc3jdUnemuTNWbsT4u8l+el53lhVVyd5SZKTqmp/kl/u7ndX1aVJbsza0nhXdvcd8xZdVTuT7NyxY8e8b2EO7owIAHBk5g3T/zLJa7r7q0lSVc/K2jrQrzvcG7v7ooOM35Dkhjk//8D37kuyb2Vl5fVD3g8AABth3jaPv/xokE6S7v5KkrMWU9LhuQARAIAxmHdm+klV9cwDZqbnfe+GMzN9bB3Y/qH1AwBgzbyB+N8m+VhV/afZ9o8nedNiSgIAgGmY9w6I762qW5O8dDb0qnW3/j7mXIAIAMAYzN2qMQvPSwvQ62nzGAftHwDA8W7eCxABAIADCNMAADDQJMO0pfEAABiDSYbp7t7X3bu2bt267FIAADiOTTJMAwDAGCztxiscXw5c+SNZW/3jYOMAAFMwyZlpPdMAAIzBJMO0nmkAAMZgkmEaAADGQJgGAICBXIDIhnObcQDgeGFmGgAABppkmLaaBwAAYzDJMG01DwAAxmCSYRoAAMZAmAYAgIGs5sEoHeltxt2WHABYBmGaSRGyAYAx0eYBAAADCdMAADDQJMO0daYBABiDSYZp60wDADAGkwzTAAAwBsI0AAAMZGk8mMPRLslnOT4A2JzMTAMAwEDCNAAADKTNg+OSOyMCABtBmIYR0GMNANOkzQMAAAYaTZiuqu+rqndX1XXLrgUAAOax0DBdVVdW1QNVdfsB4+dW1eeq6u6q2p0k3X1Pd1+8yHoAAGAjLbpn+qokb0vy3kcHqmpLkiuSvDzJ/iS3VNXe7r5zwbXApuaiSgA49hY6M93dNyf5ygHDZye5ezYT/XCSa5JcsMg6AABgEZbRM31ykvvWbe9PcnJVPbuq3pHkrKr6xYO9uap2VdWtVXXrgw8+uOhaAQDgoEazNF53P5TkkjletyfJniRZWVnpRdcFAAAHs4wwfX+SU9dtnzIbm1tV7Uyyc8eOHRtZF+g7BgCOyDLaPG5JckZVnV5VJya5MMneI9lBd+/r7l1bt25dSIEAADCPRS+Nd3WSjyV5XlXtr6qLu/uRJJcmuTHJXUmu7e47jnC/O6tqz+rq6sYXDQAAc1pom0d3X3SQ8RuS3HAU+92XZN/Kysrrh+4DAACO1mjugAgAAFMzmtU8joQLEBmLAy9YdLEiABxfJjkz7QJEAADGYJJhGgAAxkCbBxxDY2oLsaY2ABy9Sc5Ma/MAAGAMJhmmAQBgDLR5AH+G9g8AmN8kZ6a1eQAAMAaTDNMAADAGwjQAAAwkTAMAwECTDNNVtbOq9qyuri67FAAAjmOTDNMuQAQAYAwmGaYBAGAMhGkAABhImAYAgIHcAREmxh0K/RkAMB6TnJl2ASIAAGMwyTANAABjIEwDAMBAwjQAAAwkTAMAwEDCNAAADCRMAwDAQNaZBjaNg60/veh1qZf1uQAs3yRnpq0zDQDAGEwyTAMAwBgI0wAAMJAwDQAAAwnTAAAwkDANAAADCdMAADCQMA0AAAMJ0wAAMJAwDQAAA43mduJV9dQkv5bk4SQf7e73LbkkAAA4pIXOTFfVlVX1QFXdfsD4uVX1uaq6u6p2z4ZfleS67n59klcusi4AANgIi27zuCrJuesHqmpLkiuSnJfkzCQXVdWZSU5Jct/sZd9ecF0AAHDUFtrm0d03V9X2A4bPTnJ3d9+TJFV1TZILkuzPWqD+dA4R8qtqV5JdSXLaaadtfNGwyWzfff3jxu697Pyl7Ycjd+CfvT93psbvj8XbjL/rx1TLoSzjAsST89gMdLIWok9O8oEkf6eq3p5k38He3N17unulu1e2bdu22EoBAOAQRnMBYnd/M8lr53ltVe1MsnPHjh2LLQoAAA5hGTPT9yc5dd32KbOxuXX3vu7etXXr1g0tDAAAjsQywvQtSc6oqtOr6sQkFybZeyQ7qKqdVbVndXV1IQUCAMA8Fr003tVJPpbkeVW1v6ou7u5Hklya5MYkdyW5trvvOJL9mpkGAGAMFr2ax0UHGb8hyQ2L/GwAAFi0Sd5OXJsHAABjMMkwrc0DAIAxmGSYBgCAMajuXnYNg1XVg0k+f4w+7qQkXz5Gn8Wx47huXo7t5uS4bl6O7ea0mY7rX+jux90xcNJh+liqqlu7e2XZdbCxHNfNy7HdnBzXzcux3ZyOh+OqzQMAAAYSpgEAYCBhen57ll0AC+G4bl6O7ebkuG5eju3mtOmPq55pAAAYyMw0AAAMJEwDAMBAwvRhVNW5VfW5qrq7qnYvux6Gq6pTq+ojVXVnVd1RVW+cjT+rqn6rqv7X7Pszl10rR66qtlTVp6rqw7Pt06vqE7Nz9zer6sRl18iRq6pnVNV1VfU/q+quqvqrztnpq6p/NPs9fHtVXV1VT3bOTlNVXVlVD1TV7evGnvAcrTVvmR3jP6iqH1xe5RtHmD6EqtqS5Iok5yU5M8lFVXXmcqviKDyS5Oe7+8wk5yT5h7PjuTvJb3f3GUl+e7bN9LwxyV3rtn8lyZu7e0eSrya5eClVcbQuT/Jfu/svJvkrWTvGztkJq6qTk/xskpXufn6SLUkujHN2qq5Kcu4BYwc7R89Lcsbsa1eStx+jGhdKmD60s5Pc3d33dPfDSa5JcsGSa2Kg7v5Cd39y9vgbWftL+eSsHdNfn73s15P86FIKZLCqOiXJ+UneNduuJC9Nct3sJY7rBFXV1iQ/lOTdSdLdD3f31+Kc3QxOSPI9VXVCkqck+UKcs5PU3Tcn+coBwwc7Ry9I8t5e8/Ekz6iqP39MCl0gYfrQTk5y37rt/bMxJq6qtic5K8knkjynu78we+qLSZ6zrLoY7N8n+cdJvjPbfnaSr3X3I7Nt5+40nZ7kwSTvmbXwvKuqnhrn7KR19/1JfjXJH2UtRK8muS3O2c3kYOfopsxVwjTHnap6WpL3J/m57v76+ud6ba1I60VOSFW9IskD3X3bsmthw52Q5AeTvL27z0ryzRzQ0uGcnZ5Z/+wFWfvH0nOTPDWPbxNgkzgezlFh+tDuT3Lquu1TZmNMVFV9V9aC9Pu6+wOz4S89+t9Ms+8PLKs+BnlxkldW1b1Za8V6adb6bJ8x+y/kxLk7VfuT7O/uT8y2r8tauHbOTtvLkvxhdz/Y3X+a5ANZO4+ds5vHwc7RTZmrhOlDuyXJGbMrjE/M2gUSe5dcEwPN+mjfneSu7v53657am+Q1s8evSfKhY10bw3X3L3b3Kd29PWvn6H/r7p9M8pEkPzZ7meM6Qd39xST3VdXzZkN/M8mdcc5O3R8lOaeqnjL7vfzocXXObh4HO0f3Jvmp2aoe5yRZXdcOMlnugHgYVfUjWevH3JLkyu5+03IrYqiq+utJ/nuSz+ax3tp/mrW+6WuTnJbk80l+orsPvJiCCaiqlyT5he5+RVV9X9Zmqp+V5FNJXt3d31pieQxQVS/I2oWlJya5J8lrszYR5JydsKr6F0n+btZWWfpUkr+ftd5Z5+zEVNXVSV6S5KQkX0ryy0k+mCc4R2f/eHpb1tp6/jjJa7v71iWUvaGEaQAAGEibBwAADCRMAwDAQMI0AAAMJEwDAMBAwjQAAAwkTAMAwEDCNAAADCRMA4xcVW2vqruq6p1VdUdV3VRV3/MEr9tWVe+vqltmXy+ejX+oqn5q9vhnqup9s8cXVdVnq+r2qvqVdfv5P1X1pqr6TFV9vKqeMxv/8dlrP1NVNx+bnx5g3Ny0BWDkqmp7kruTrHT3p6vq2iR7u/s/HPC6/5jk17r7f1TVaUlu7O4fmIXh383a3QPfneScJE9O8vEkL0zy1SQ3JXlLd3+wqjrJK7t7X1X96yRf7+5/VVWfTXJud99fVc/o7q8di58fYMxOWHYBAMzlD7v707PHtyXZ/gSveVmSM9fu2JskeXpVPa27v1RVv5TkI0n+9uy2vhck+Wh3P5gks9nqH8rabYAfTvLhdZ/18tnj301y1SzMf2ADfzaAyRKmAabhW+sefzvJ49o8sta6d053/8kTPPeXkjyU5LlzfNaf9mP/bfntzP6u6O5LqupFSc5PcltVvbC7H5r3BwDYjPRMA2weNyV5w6MbVfWC2fezk5yX5Kwkv1BVpyf5/SR/o6pOqqotSS5K8juH2nlVfX93f6K7fynJg0lOXchPATAhwjTA5vGzSVaq6g+q6s4kl1TVdyd5Z5LXdff/TvLzSa5M8sUku7PW+vGZJLd194cOs/9/8+gFi0l+b/Y+gOOaCxABAGAgM9MAADCQMA0AAAMJ0wAAMJAwDQAAAwnTAAAwkDANAAADCdMAADDQ/wODYHY4swDAswAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.figure(figsize=(12,4))\n",
    "plt.bar(*np.unique(n_exons,return_counts=True))\n",
    "plt.xlabel('n exons');\n",
    "plt.yscale('log')\n",
    "plt.ylabel('counts');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4c53cbf-11ae-4340-bce4-e590803bba0d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
